presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayPositionWithIndexFunction.java

```diff
@@ -0,0 +1,146 @@
+            @OperatorDependency(operator = EQUAL, argumentTypes = {"T", "T"}) MethodHandle equalMethodHandle,
+            @SqlType("array(T)") Block array,
+            @SqlType("T") double element,
+            @SqlType(StandardTypes.BIGINT) long instance)
+    {
+        return arrayPositionWithIndexInternal(type::getDouble, equalMethodHandle, array, element, instance);
+    }
+
+    @TypeParameter("T")
+    @SqlType(StandardTypes.BIGINT)
+    public static long arrayPositionWithIndex(
+            @TypeParameter("T") Type type,
+            @OperatorDependency(operator = EQUAL, argumentTypes = {"T", "T"}) MethodHandle equalMethodHandle,
+            @SqlType("array(T)") Block array,
+            @SqlType("T") Slice element,
+            @SqlType(StandardTypes.BIGINT) long instance)
+    {
+        return arrayPositionWithIndexInternal(type::getSlice, equalMethodHandle, array, element, instance);
+    }
+
+    @TypeParameter("T")
+    @SqlType(StandardTypes.BIGINT)
+    public static long arrayPositionWithIndex(
+            @TypeParameter("T") Type type,
+            @OperatorDependency(operator = EQUAL, argumentTypes = {"T", "T"}) MethodHandle equalMethodHandle,
+            @SqlType("array(T)") Block array,
+            @SqlType("T") Block element,
+            @SqlType(StandardTypes.BIGINT) long instance)
+    {
+        return arrayPositionWithIndexInternal(type::getObject, equalMethodHandle, array, element, instance);
+    }
+
+    protected static long arrayPositionWithIndexInternal(TypeValueExtractor extractor, MethodHandle equalMethodHandle, Block array, Object element, long instance)
+    {
+        int size = array.getPositionCount();
+        int instancesFound = 0;
+
+        if (instance == 0) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "array_position cannot take a 0-valued instance argument.");
+        }
+
+        int startIndex = instance > 0 ? 0 : size - 1;
+        int stopIndex = instance > 0 ? size : -1;
+        int stepSize = instance > 0 ? 1 : -1;
+        instance = Math.abs(instance);
+
+        for (int i = startIndex; i != stopIndex; i += stepSize) {
+            if (!array.isNull(i)) {
+                try {
+                    Object arrayValue = extractor.extract(array, i);
+                    Boolean result = (Boolean) equalMethodHandle.invoke(arrayValue, element);
+                    checkNotIndeterminate(result);
+                    if (result) {
+                        instancesFound++;
+                        if (instancesFound == instance) {
+                            return i + 1; // result is 1-based (instead of 0)
+                        }
+                    }
+                }
+                catch (Throwable t) {
+                    throw internalError(t);
+                }
+            }
+        }
+        return 0;
+    }
+
+    protected interface TypeValueExtractor
+    {
+        Object extract(Block array, int index);
```

Comment:
We had 5 separate implementations so we can avoid boxing primitive types and avoid invoke dynamic. So Let's keep the 5 separate implementations. Performance is more important in this case.

Reply:
Okay. I will split this to five implementations.

Reply:
I'm not entirely sure about the 5 specializations. Those micro benchmarks may be misleading.

Reply:
If there's doubt about the necessity we can reconsider this in a separate PR. Whether the performance is acceptable without specialization is not in scope of this PR.

Reply:
@rongrong Okay. I have now kept five implementations.